src/grid/cubic.py

```diff
@@ -0,0 +1,575 @@
+# GRID is a numerical integration module for quantum chemistry.
+#
+# Copyright (C) 2011-2019 The GRID Development Team
+#
+# This file is part of GRID.
+#
+# GRID is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 3
+# of the License, or (at your option) any later version.
+#
+# GRID is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, see <http://www.gnu.org/licenses/>
+# --
+r"""3D Integration Grid."""
+
+from grid.basegrid import Grid, OneDGrid
+
+import numpy as np
+
+from scipy.interpolate import CubicSpline
+
+from sympy import symbols
+from sympy.functions.combinatorial.numbers import bell
+
+
+class _CubicGrid(Grid):
+    def __init__(self, points, weights, shape):
+        r"""
+        Construct the CubicGrid class, i.e. each point is specified by (i, j, k) indices.
+
+        Parameters
+        ----------
+        points : np.ndarray(M, 3)
+            The 3D points of the cubic grid.
+        weights : np.ndarray(M*3)
+            The weights corresponding to each cubic grid point.  It moves in the following
+            order (x, y, z).
+        shape : np.ndarray(3)
+            The number of points in each of the three (x, y, z) directions.
+
+        """
+        if len(shape) != 3:
+            raise ValueError(
+                "Shape (length {0}) should have length three.".format(len(shape))
+            )
+        self._shape = shape
+        super().__init__(points, weights)
+
+    @property
+    def shape(self):
+        r"""Return number of points in each direction."""
+        return self._shape
+
+    def interpolate_function(
```

Comment:
Thanks @Ali-Tehrani for including the interpolation functionality as well. I wonder whether this should be turned into a private method or we should add an interpolation function to other grids as well? I am also wondering whether this should just return the splines? This is because we have an interpolation module, but if I am not wrong, it assumes the given grid is an atom-centered spherical grid and returns spline (to be used for example for an ODE solver)... What are your thoughts @Ali-Tehrani?

What do you think @tczorro? Can you please share a code snippet to show how the refactored interpolation/evaluation module works?

Reply:
Don't think it's worthwhile to add interpolation function to 1DGrids, because of how easy it is (1-2 lines of code). Not sure of other grids.
There wouldn't be universal spline that would work anywhere because interpolation of the x-coordinate would depend on the y-coordinate you want to interpolate etc.  Since it's nested interpolation, the only cubic splines that is used everywhere is a list of cubic spline for each point in the z-axis.  After that, it will use each of the cubic spline in the z-axis to construct another cubic spline in the y-axis that will depend on the point you want to interpolate. But not entirely certain, need to double check.
Don't know much about having a private method because I don't know the use-case here.